005-1-locations-on-stack.md

```diff
@@ -0,0 +1,590 @@
+> The DW_OP_push_object_address operation pushes the location of the
+> object currently being evaluated as part of evaluation of a user
+> presented expression. This object may correspond to an independent
+> variable described by its own debugging information entry or it may be a
+> component of an array, structure, or class whose address has been
+> dynamically determined by an earlier step during user expression
+> evaluation. This operator provides explicit functionality (especially
+> for arrays involving descriptors) that is analogous to the implicit push
+> of the base address of a structure prior to evaluation of a
+> DW_AT_data_member_location to access a data member of a structure. For
+> an example, see Appendix D.2 on page 304.
+
+
+### Section 2.5.2.1 Memory Locations
+
+Insert the following (adapted from Section 2.6.1.1.2):
+
+> A memory location represents the location of a piece or all of an
+> object or other entity in memory. On architectures that support
+> multiple address spaces, a memory location contains a component that
+> identifies the address space (which may be provided by the
+> `DW_OP_xderef` operation). A memory location is considered
+> _unbounded_, as the size of the location storage is only implied by
+> the type of object stored at that location.
+> 
+> In contexts that expect a location, a value of the generic type
+> may be implicitly converted to a memory location in the default
+> address space.
+> 
+> The following operations push memory locations onto the DWARF stack:
+> 
+> 1. `DW_OP_addr` [moved from Section 2.5.1.1]  
+> The `DW_OP_addr` operation has a single operand that encodes a
+> machine address and whose size is the size of an address on the
+> target machine. The value of this operand is treated as an address
+> in the default address space and pushed onto the stack.
+> 
+> 2. `DW_OP_addrx` [moved from Section 2.5.1.1]  
+> The `DW_OP_addrx` operation has a single operand that encodes an
+> unsigned LEB128 value, which is a zero-based index into the `.debug\_addr`
+> section, where a machine address is stored. This index is relative to the
+> value of the `DW_AT_addr_base` attribute of the associated compilation
+> unit. The address obtained is treated as an address in the default address
+> space and pushed onto the stack.
+>
+> 3. `DW_OP_form_tls_address`... [moved from Section 2.5.1.3]
+> 
+> 4. `DW_OP_call_frame_cfa`... [moved from Section 2.5.1.3]
+
+
+### Section 2.5.2.2 Register Locations
+
+Move the contents of Section 2.6.1.1.3 here, replacing the term
+"location description" with "location" throughout.
+
+[page 39] Remove the non-normative text:
+
+> A register location description must stand alone as the entire
+> description of an object or a piece of an object.
+
+
+### Section 2.5.2.3 Implicit Locations
+
+Move the contents of Section 2.6.1.1.4 here, replacing the term
+"location description" with "location" throughout.
```

Comment:
There is one sentence: "DWARF location descriptions are intended to yield the location of a value rather than the value itself." I think it should not be replaced in this instance.

Don't we need to edit the following text for `DW_OP_stack_value` more:
"In this form of location description, the DWARF expression
represents the actual value of the object, rather than its location."

to something like:
"In this form of location, the value at the top of the DWARF expression
stack represents the actual value of the object, rather than its location."

There is also
"DW_OP_stack_value operation terminates the expression."
Should this limitation stay? Why not just remove it to have more flexibility? `DW_OP_stack_value` is essentially the same as `DW_OP_implicit_value`, except that the contents come from the top of the stack instead of the operand. An implicit value does not say that it should terminate the expression.